src/depends/NAT/nat.cpp

```diff
@@ -0,0 +1,176 @@
+/*
+	This file is part of cpp-ethereum.
+	cpp-ethereum is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+	cpp-ethereum is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+	You should have received a copy of the GNU General Public License
+	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/// @file
+/// UPnP port forwarding support.
+
+#include "nat.h"
+// #include "libUtils/Logger.h"
+#include <arpa/inet.h> //INET6_ADDRSTRLEN
+#include <cstring>
+#include <iostream>
+#include <miniupnpc/miniupnpc.h>
+#include <miniupnpc/upnpcommands.h>
+#include <miniupnpc/upnperrors.h>
+
+using namespace std;
+
+namespace
+{
+    const int DISCOVERY_TIME_OUT = 2000;
+    const int TTL  = 2;   
+}
+
+NAT::NAT()
+{
+    m_urls = make_shared<UPNPUrls>();
+    m_data = make_shared<IGDdatas>();
+
+    memset(m_urls.get(), 0, sizeof(struct UPNPUrls));
+    memset(m_data.get(), 0, sizeof(struct IGDdatas));
+
+    m_initialized = false;
+
```

Comment:
At this point, I would "stop" the constructor and start an `init()` method. `upnpDiscover` and `UPNP_GetValidIGD` are both very heavy and I think it would be better if this was returning error codes. Overall people recommend constructors to be relatively "light" and if they end up "heavy" turn them into `init()` methods. By using the `m_initialized` member, you're anyway halfway there.